test/http/http-client.test.js

```diff
@@ -0,0 +1,209 @@
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const should = require('should');
+
+const { importFile, getTestOptions } = require('../testutils');
+const MockRequest = require('../mock-request');
+
+const HttpClient = importFile('http/http-client');
+const HttpRequest = importFile('http/http-request');
+const DirectBinaryUploadOptions = importFile('direct-binary-upload-options');
+const HttpResult = importFile('http-result');
+
+describe('HTTP Client Tests', function() {
+    let httpResult;
+    let options;
+    let uploadOptions;
+    let httpClient;
+
+    beforeEach(function() {
+        MockRequest.reset();
+        options = getTestOptions();
+        uploadOptions = new DirectBinaryUploadOptions()
+            .withHttpRetryDelay(10);
+        httpClient = new HttpClient(options, uploadOptions);
+        httpResult = new HttpResult(options, uploadOptions);
+    });
+
+    function createHttpRequest(url) {
+        return new HttpRequest(options, url);
+    }
+
+    it('http request submit smoke test', async function() {
+        MockRequest.onGet('http://myunittest.com').reply(200, 'hello world!');
+        const response = await httpClient.submit(createHttpRequest('http://myunittest.com'));
+
+        should(response).be.ok();
+        should(response.getStatusCode()).be.exactly(200);
+        should(response.getData()).be.exactly('hello world!');
+        should(response.getElapsedTime() !== undefined).be.ok();
+    });
+
+    it('test submit retry', async function() {
+        const url = 'http://myretryunittest';
+        MockRequest.onGet(url).replyOnce(500);
+        MockRequest.onGet(url).reply(200, 'retried!');
+
+        const response = await httpClient.submit(createHttpRequest(url), httpResult);
+        should(response).be.ok();
+
+        should(MockRequest.history.get.length).be.exactly(2);
+        should(httpResult.getRetryErrors().length).be.exactly(1);
+    });
+
+    it('test submit retry all fail', async function() {
+        const url = 'http://myfailingretryunittest';
+        MockRequest.onGet(url).reply(500);
+
+        let threw = false;
+        try {
+            await httpClient.submit(createHttpRequest(url));
+        } catch (e) {
+            threw = true;
+        }
+        should(MockRequest.history.get.length).be.exactly(3);
+        should(threw).be.ok();
+    });
+
+    it('test submit retry not retryable error', async function() {
+        const url = 'http://mynotfoundunittest';
+
+        let threw = false;
+        try {
+            await httpClient.submit(createHttpRequest(url));
+        } catch (e) {
+            threw = true;
+        }
+        should(threw).be.ok();
```

Comment:
Something based on `assertionRejected` may be better.

Reply:
Do you have any examples of this?

Reply:
- For should, I don't have examples except the doc: https://shouldjs.github.io/#assertion-rejected or https://shouldjs.github.io/#assertion-rejectedwith.
- For expect (aio uses expect): https://github.com/adobe/aio-cli-plugin-runtime/blob/master/test/commands/runtime/activation/get.test.js#L117
- For assert: https://github.com/adobe/asset-compute-sdk/blob/master/test/shellscript.test.js#L274

Note that `assert` is build-in nodejs, so you don't need extra/additional dependencies to test it.